src/libAtomVM/term.c

```diff
@@ -379,3 +379,57 @@ int term_compare(term t, term other, Context *ctx)
 
     return result;
 }
+
+term term_alloc_refc_binary(Context *ctx, const void *data, size_t size, unsigned is_const)
+{
+    term *boxed_value = memory_heap_alloc(ctx, TERM_BOXED_REFC_BINARY_SIZE);
+    boxed_value[0] = ((TERM_BOXED_REFC_BINARY_SIZE - 1) << 6) | TERM_BOXED_REFC_BINARY;
+    boxed_value[1] = (term) size;
+    boxed_value[2] = (term) is_const ? 0x01 : 0x00;
+    term ret = ((term) boxed_value) | TERM_BOXED_VALUE_TAG;
+    if (is_const) {
+        boxed_value[3] = (term) data;
+        boxed_value[4] = term_nil(); // mso_list is not used
+        boxed_value[5] = term_nil(); // for const binaries
+    } else {
+        void *refc = memory_create_refc(size, data);
+        boxed_value[3] = (term) refc;
+        ctx->mso_list = term_list_init_prepend(boxed_value + 4, ret, ctx->mso_list);
+        list_append(&ctx->global->refc_binaries, (struct ListHead *) refc);
+    }
+    return ret;
+}
+
+static term find_finary(term binary_or_state) {
+    if (term_is_match_state(binary_or_state)) {
+        find_finary(term_get_match_state_binary(binary_or_state));
+    } else {
+        return binary_or_state;
+    }
+}
+
+term term_alloc_bin_match_state(term binary_or_state, Context *ctx)
+{
+    term *boxed_match_state = memory_heap_alloc(ctx, TERM_BOXED_BIN_MATCH_STATE_SIZE);
+
+    boxed_match_state[0] = ((TERM_BOXED_BIN_MATCH_STATE_SIZE - 1) << 6) | TERM_BOXED_BIN_MATCH_STATE;
+    if (term_is_match_state(binary_or_state)) {
+        boxed_match_state[1] = term_get_match_state_binary(binary_or_state);
+        boxed_match_state[2] = term_get_match_state_offset(binary_or_state);
+        boxed_match_state[3] = term_get_match_state_offset(binary_or_state);
+    } else {
+        boxed_match_state[1] = binary_or_state;
+        boxed_match_state[2] = 0;
+        boxed_match_state[3] = 0;
+    }
+    term binary = find_finary(binary_or_state);
+    term *boxed_value = term_to_term_ptr(binary);
+    if (term_is_refc_binary(binary) && !term_refc_binary_is_const(binary)) {
+        memory_increment_recf_refcount(boxed_value[3]);
+        ctx->mso_list = term_list_init_prepend(boxed_match_state + 4, binary, ctx->mso_list);
+    } else {
+        boxed_match_state[4] = term_nil(); // mso_list is not used
+        boxed_match_state[5] = term_nil(); // for const binaries
```

Comment:
we are using 8 more bytes for each match state, is it possible to avoid this?

Reply:
If you could convince me that a GC can't occur between the time a match state is created and the time a match state is created (`OP_BS_START_MATCH*`) and the time it is converted to a binary (`OP_BS_GET_TAIL`), then perhaps creation of a match state does not need to increment the reference count.  But I couldn't find any guarantee that this would not be the case.  Do you know?

Reply:
I need to think about this

Reply:
Sounds good.  I was thinking maybe we could just increment the reference count on the _referenced_ binary, but the problem with that is that it would not be tracked in the MSO list, so unless you were willing to scan the entire source heap for unmarked match states (something that the MSO list is designed to avoid), you'd end up with reference counts that will never decrement to 0.

Reply:
still not sure here.
During GC matches are parsed as well:
from (master) memory.c:
```c
305                 case TERM_BOXED_BIN_MATCH_STATE: {
306                     TRACE("- Found bin match state.\n");
307                     ptr[1] = memory_shallow_copy_term(ptr[1], &new_heap, move);
308                     break;
309                 }
```
so from the garbage collector point of view a match state is just a boxed term like others, such as tuples (and we are not treating them in a different way).

I think that reference counter should be incremented only when creating a new boxed term (having type refc binary), but here it is never created such a similar boxed term, but an already existing one is always referenced (and stored in boxed_match_state[1]).

am I missing any relevant point?

Reply:
I think you are right -- we do not externalize match states.  WE will need to tread carefully when implementing sub-binaries, but we don't need to track these in the MSO list.  Code has been remove.